File: src/mining/here.ts
// The stationary half of dist/mining.js: no scan, no walk, no block map and no respawn wait. It can
// leave all of that out because dig.ts answers the cursor with yourself, so a swing is aimed by
// where the character stands rather than by naming a tile.
//
// It does not move at all, which is why the smelt is smeltHere rather than smeltAll: a beetle that
// has wandered off is a pass skipped, not a walk taken.
import { describeItem } from '../lib/entity.js';
import { runHarvest, type Handled, type Interlude } from '../lib/harvest.js';
import { createStallWatch } from '../lib/loop.js';
import { overweight } from '../lib/weight.js';
import {
  LOG_EVERY,
  MAX_CYCLES,
  MAX_NO_CURSOR,
  MAX_THROTTLED,
  MAX_UNKNOWN,
  PICK_BEETLE,
  STALL_STOP,
  STALL_WARN,
  STEP_DELAY,
  THROTTLE_BACKOFF,
  THROTTLE_BACKOFF_MAX,
} from './config.js';
import { digOnce } from './dig.js';
import { stopReason } from './guards.js';
import { heartbeat } from './heartbeat.js';
import { dismount } from './mount.js';
import { groupOres, oreTotal, waitForOre } from './ore.js';
import { equipPickaxe, pickaxeSerial, rememberPickaxe } from './pickaxe.js';
import { waitOutSave } from './save.js';
import { pickBeetle, retryUnsmeltable, smeltHere } from './smelt.js';
import { watchForTrouble } from './threat.js';

rememberPickaxe(player.equippedItems.oneHanded);

const tooHeavy = (): boolean => overweight();

const WORKED_OUT = 'the spot is worked out';

log(`mine-here: ${oreTotal()} ore in the pack to start, at ${player.x},${player.y}`);

// A run that stops on its first cycle looks from the outside like one that never started. The
// position matters more here than in dist/mining.js: where the character stands is the whole premise.
// Read before the dismount below, or the mounted half of it always answers no.
log(
  `mine-here: mounted ${player.equippedItems.mount ? 'yes' : 'no'}, ` +
    `hand ${describeItem(player.equippedItems.oneHanded)}, ` +
    `weight ${player.weight}/${player.weightMax}`,
);

// Before the cursor, so the beetle you click is one standing next to you rather than the one you are
// sitting on
const afoot = dismount();

if (PICK_BEETLE) {
  pickBeetle();
}

// A pack that arrives full has no room for the first swing's ore, and the loop would find that out by
// destroying it. Smelting only once off the mount: a smelt aimed at the beetle you ride is silent,
// and three silent passes write the hue off before the run has started.
groupOres();

if (afoot && tooHeavy()) {
  smeltHere();
}

let oreBefore = 0;

// A real ending here in a way it is not in dist/mining.js, which empties the pack every time it walks
// on: a beetle out of range is a beetle this run will not walk to. Read the ending it leads to as
// "put the beetle next to me" rather than as a bug.
const smeltForRoom = (): Interlude => {
  if (!tooHeavy()) {
    return undefined;
  }

  const before = oreTotal();

  // Unconditional: asking tooHeavy() a second time inside a helper is how a live run once stopped
  // overweight without ever having tried.
  groupOres();
  smeltHere();

  // Ore leaving the pack is the proof a smelt landed, not the weight going down: the client can
  // still be reporting its pre-smelt figure when the pack diff has confirmed the conversion.
  if (oreTotal() < before) {
    return { phase: 'smelting' };
  }

  // Likeliest reason is a hue given up on earlier - a beetle briefly out of range looks exactly like
  // an ore that cannot be worked. False once a retry has been spent without freeing any ore, which is
  // what stops this cycling through 'smelting' until the stall watchdog.
  if (retryUnsmeltable()) {
    return { phase: 'smelting' };
  }

  // The last thing tried rather than a second helping of the first: the retry above has just reopened
  // the hues written off, so this pass is the one that can act on them.
  groupOres();
  smeltHere();

  if (oreTotal() < before) {
    return { phase: 'smelting' };
  }

  return {
    stop:
      `overweight (${player.weight}/${player.weightMax}) with ${oreTotal()} ore left, ` +
      'and smelting freed nothing - the beetle has to be standing next to you',
  };
};

const handle = (outcome: string): Handled => {
  switch (outcome) {
    // The two ways the shard says there is nothing left. In dist/mining.js they differ by scope,
    // which is what decides where to walk next; there is no next here, so both mean the same thing.
    case 'empty':
    case 'nothingNearby':
      groupOres();
      smeltHere();

      return { stop: WORKED_OUT };

    // About a swing that named no tile, so the shard is saying this spot is not mineable at all.
    // Nothing to ban and nowhere to walk, so it is an ending.
    case 'notOre':
      return { stop: 'nothing here can be mined' };

    // Range and line of sight, neither of which can be answered by moving - the one thing this
    // script does not do. Named separately because one is a shard that wanted a tile after all and
    // the other is something in the way.
    case 'tooFar':
      return { stop: 'the shard says the ore is out of reach from where you are standing' };

    case 'notSeen':
      return { stop: 'the shard cannot see the ore from where you are standing' };

    // The ore this swing produced was destroyed rather than dropped, so swinging again destroys
    // more. A full pack is a container at its item cap, so consolidating is the fix: forty piles of
    // one become one pile of forty. If weight is the real problem, the next cycle smelts.
    case 'packFull':
      log('mine-here: pack is full, consolidating before the next swing');
      groupOres();

      return {};

    default:
      return undefined;
  }
};

runHarvest({
  prefix: 'mine-here',
  landed: 'dug',
  toolName: 'pickaxe',

  stopReason,
  watch: watchForTrouble('mine-here'),
  equipTool: equipPickaxe,
  ready: () => (dismount() ? undefined : 'could not get off the mount'),
  relieve: smeltForRoom,
  waitOutSave,

  harvest: () => {
    oreBefore = oreTotal();

    return digOnce(pickaxeSerial());
  },

  // Ore arrives as a new pile rather than joining the one already there. Grouping every swing keeps
  // the pack at one pile per hue, so the item cap is never approached by pile count alone - packFull
  // destroys the ore of the swing that hits it.
  onLanded: () => {
    waitForOre(oreBefore);
    groupOres();
  },

  handle,

  progress: (mined) => `${mined} swings, ${oreTotal()} ore`,

  finish: (mined, reason) => {
    // The worked-out ending has already smelted by the time it gets here, and every other ending is
    // something being wrong - what is in the pack then keeps perfectly well as ore.
    groupOres();

    if (tooHeavy()) {
      smeltHere();
    }

    // A worked-out ending with no swing landed did not work anything out: it was standing somewhere
    // with no ore in it, and the stop reason alone reads like the end of a good run.
    if (reason === WORKED_OUT && mined === 0) {
      log(
        'mine-here: no swing ever landed - the character is probably not standing next to a vein',
      );
    }

    // Swings rather than an ore delta: smelted ore has left the pack, so the pack cannot total the
    // run
    log(`mine-here: ${mined} swings, ${oreTotal()} ore still in the pack`);
  },

  stall: createStallWatch({
    prefix: 'mine-here',
    without: 'cycles without a swing landing',
    warnAt: STALL_WARN,
    stopAt: STALL_STOP,
    heartbeat,
  }),

  timings: {
    stepDelay: STEP_DELAY,
    maxCycles: MAX_CYCLES,
    maxUnknown: MAX_UNKNOWN,
    maxThrottled: MAX_THROTTLED,
    maxNoCursor: MAX_NO_CURSOR,
    logEvery: LOG_EVERY,
    throttleBackoff: THROTTLE_BACKOFF,
    throttleBackoffMax: THROTTLE_BACKOFF_MAX,
  },
});

File: legion/src/mining/here.py
import API

from mining.config import (LOG_EVERY, MAX_CYCLES, MAX_NO_CURSOR, MAX_NO_TOOL, MAX_THROTTLED,
                           MAX_UNKNOWN, ORE_SETTLE_POLL, ORE_SETTLE_TIMEOUT, OUTCOME_TEXT,
                           PICK_BEETLE, STEP_DELAY, THROTTLE_BACKOFF, THROTTLE_BACKOFF_MAX)
from mining.dig import Digger
from mining.relieve import Relief
from mining.run import Run
from uo.loop import backoff_for
from uo.weight import too_heavy

run = Run("mine-here")

log = run.log
heartbeat = run.heartbeat
stall = run.stall
stop_reason = run.stop_reason
saves = run.saves
pickaxe = run.pickaxe
ore = run.ore
combiner = run.combiner
beetle = run.beetle
smelter = run.smelter
threat = run.threat
gathered = run.gathered
get_off_the_mount = run.get_off_the_mount
say_where_we_stand = run.say_where_we_stand

WORKED_OUT = "the spot is worked out"

# No pathfind to cancel: this run never walks, so a cancel here would only fight the player
digger = Digger(ore, OUTCOME_TEXT, run.dig_config, log, False)
relief = Relief(ore, combiner, smelter, saves, beetle.in_range,
                " - the beetle has to be standing next to you", log)

say_where_we_stand()

# Before the cursor, so the beetle you click is one standing next to you rather than the one you are
# sitting on
afoot = get_off_the_mount()

if PICK_BEETLE:
    beetle.pick()

# A pack that arrives full has no room for the first swing's ore. Smelting only once off the mount:
# a smelt aimed at the beetle you ride is silent, and three silent passes write the hue off.
combiner.group()

if afoot and too_heavy():
    relief.smelt()

stop = None
tally = 0
fails = 0
unknown = 0
throttled = 0
no_cursor = 0
no_tool = 0
reported = 0
cycle = 0


def end_cycle(phase):
    global stop

    stall.end_cycle(phase, cycle, tally)

    if stop is None:
        stop = stall.reason()


try:
    while stop is None and cycle < MAX_CYCLES:
        cycle += 1

        stop = stop_reason()

        if stop is not None:
            break

        # Everything below reads a frozen shard as its own failure: a smelt that converts nothing is
        # ore that cannot be worked
        if saves.is_saving():
            saves.wait_out()
            unknown = 0
            throttled = 0
            stall.progressed()
            end_cycle("saving")
            continue

        threat.look()

        # Asked every cycle, so a remount costs a single cycle instead of the rest of the run
        if not get_off_the_mount():
            stop = "could not get off the mount"
            break

        if not pickaxe.equip():
            no_tool += 1

            if no_tool >= MAX_NO_TOOL:
                stop = "no pickaxe"
                break

            log("no pickaxe (%d/%d), looking again" % (no_tool, MAX_NO_TOOL))
            end_cycle("no tool")
            API.Pause(backoff_for(no_tool, THROTTLE_BACKOFF, THROTTLE_BACKOFF_MAX))
            continue

        no_tool = 0

        relieved = relief.smelt_for_room()

        if relieved is not None:
            if isinstance(relieved, dict):
                stop = relieved["stop"]
                break

            end_cycle(relieved)
            API.Pause(STEP_DELAY)
            continue

        value = gathered.settle()
        before = gathered.before_swing()
        ore_before = ore.total()
        outcome = digger.dig_once(pickaxe.serial())

        if outcome == "dug":
            tally += 1
            unknown = 0
            throttled = 0
            stall.progressed()

            # Ore arrives as a new pile after the sentence that announced it, so grouping every
            # swing keeps the pack at one pile per metal and the item cap out of reach
            ore.wait_for_ore(ore_before, ORE_SETTLE_TIMEOUT, ORE_SETTLE_POLL)
            combiner.group()
            gathered.after_swing(value, "dug", before)

        elif outcome == "failed":
            tally += 1
            fails += 1
            unknown = 0
            throttled = 0
            stall.progressed()
            gathered.after_swing(value, "failed", before)

        elif outcome == "wornOut":
            log("pickaxe worn out, swapping")
            unknown = 0

        elif outcome == "saving":
            saves.wait_out()
            unknown = 0
            throttled = 0
            stall.progressed()

        elif outcome == "throttled":
            throttled += 1

            # A refusal is a read outcome, so it clears the unreadable count: left standing, a shard
            # alternating refusals with silence ends the run on MAX_UNKNOWN
            unknown = 0
            log("shard says wait (%d/%d), backing off" % (throttled, MAX_THROTTLED))
            API.Pause(backoff_for(throttled, THROTTLE_BACKOFF, THROTTLE_BACKOFF_MAX))

            if throttled >= MAX_THROTTLED:
                stop = "the shard kept refusing the swing"

        elif outcome == "noCursor":
            no_cursor += 1
            log("no target cursor (%d/%d), backing off" % (no_cursor, MAX_NO_CURSOR))
            API.Pause(backoff_for(no_cursor, THROTTLE_BACKOFF, THROTTLE_BACKOFF_MAX))

            if no_cursor >= MAX_NO_CURSOR:
                stop = "the shard never opened a target cursor"

        # One branch, not two: the pair differ by scope, and scope only matters to a run with
        # somewhere else to walk
        elif outcome == "empty" or outcome == "nothingNearby":
            unknown = 0
            combiner.group()
            relief.smelt()
            stop = WORKED_OUT

        # No art to ban and nowhere to walk, so it is an ending
        elif outcome == "notOre":
            unknown = 0
            stop = "nothing here can be mined"

        # Neither is answerable by moving, and this run does not move
        elif outcome == "tooFar":
            unknown = 0
            stop = "the shard says the ore is out of reach from where you are standing"

        elif outcome == "notSeen":
            unknown = 0
            stop = "the shard cannot see the ore from where you are standing"

        # The ore this swing produced was destroyed rather than dropped, so a full pack is answered
        # by consolidating: forty piles of one become one pile of forty
        elif outcome == "packFull":
            unknown = 0
            log("pack is full, consolidating before the next swing")
            combiner.group()

        else:
            unknown += 1
            log("unreadable outcome (%d/%d), check OUTCOME_TEXT" % (unknown, MAX_UNKNOWN))

        # Done here rather than inside each branch the way unknown is: every branch but one clears
        # it, and one added later would have to remember to
        if outcome != "noCursor":
            no_cursor = 0

        if unknown >= MAX_UNKNOWN:
            stop = "%d unreadable outcomes in a row" % MAX_UNKNOWN
            break

        if tally >= reported + LOG_EVERY:
            reported = tally
            log(
                "%d swings, %d ore, %d/%d"
                % (tally, ore.total(), API.Player.Weight, API.Player.WeightMax)
            )

        end_cycle(outcome if outcome is not None else "unknown")
        API.Pause(STEP_DELAY)
except Exception as error:
    # The stop button lands here as well, and the client waits for it to unwind the thread
    if API.StopRequested:
        raise

    # Nothing else catches: a throw out of a client call used to end the run with no line at all
    if stop is None:
        stop = "threw - %s" % error

reason = stop or "hit the %d working cycle backstop" % MAX_CYCLES

# Smelted only if the run is ending over the limit: what is in the pack is a few swings' worth
combiner.group()

if too_heavy():
    relief.smelt()

gathered.settle()

# Swings rather than an ore delta: smelted ore has left the pack, so the pack cannot total the run
log("%d swings, %d failed, %d ore still in the pack" % (tally, fails, ore.total()))

if reason == WORKED_OUT and tally == 0:
    log("no swing ever landed - the character is probably not standing next to a vein")
log("stopping - %s" % reason)
API.Stop()
